pkg/agent/memberlist/cluster.go

```diff
@@ -0,0 +1,493 @@
+	n, err := c.mList.Join(c.members().List())
+	if err != nil {
+		klog.ErrorS(err, "Join cluster failed", "num", n)
+	}
+	if err := c.initConsistentHash(); err != nil {
+		klog.ErrorS(err, "Init consistentHashMap error")
+		return
+	}
+
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+
+	for {
+		select {
+		case <-stopCh:
+			return
+		case nodeEvent := <-c.nodeEventsCh:
+			c.handleClusterNodeEvents(&nodeEvent)
+		}
+	}
+}
+
+func (c *Cluster) initConsistentHash() error {
+	node, err := c.nodeLister.Get(c.nodeName)
+	if err != nil {
+		return err
+	}
+	eips := c.filterEIPsFromNodeLabels(node)
+	c.enqueueEIPs(eips)
+	return nil
+}
+
+func (c *Cluster) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Cluster) processNextWorkItem() bool {
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(obj)
+
+	// We expect strings (externalIPPool name) to come off the workqueue.
+	if key, ok := obj.(string); !ok {
+		// As the item in the workqueue is actually invalid, we call Forget here else we'd
+		// go into a loop of attempting to process a work item that is invalid.
+		// This should not happen.
+		c.queue.Forget(obj)
+		klog.Errorf("Expected string in work queue but got %#v", obj)
+		return true
+	} else if err := c.syncConsistenHash(key); err == nil {
+		// If no error occurs we Forget this item so it does not get queued again until
+		// another change happens.
+		c.queue.Forget(key)
+	} else {
+		// Put the item back on the workqueue to handle any transient errors.
+		c.queue.AddRateLimited(key)
+		klog.ErrorS(err, "Syncing failed, requeue", "eip", key)
+	}
+	return true
+}
+
+func (c *Cluster) syncConsistenHash(eipName string) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).InfoS("Finished syncing consistentHash", "eip", eipName, "durationTime", time.Since(startTime))
+	}()
+
+	c.consistentHashRWMutex.Lock()
+	defer c.consistentHashRWMutex.Unlock()
+	eip, err := c.externalIPPoolLister.Get(eipName)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			delete(c.consistentHashMap, eipName)
+			return nil
+		}
+		return err
+	}
+
+	if err := c.syncConsistentHashByEIP(eip); err != nil {
+		klog.ErrorS(err, "Sync consistentHash by EIP failed", "EIP", eip.Name)
+		return err
+	}
+
+	return nil
+}
+
+func newNodeConsistentHashMap() *consistenthash.Map {
+	return consistenthash.New(defaultVirtualNodeReplicas, defaultHashFn)
+}
+
+// syncConsistentHashByEIP refresh the consistentHashMap
+func (c *Cluster) syncConsistentHashByEIP(eip *v1alpha2.ExternalIPPool) error {
+	nodeSel, err := metav1.LabelSelectorAsSelector(&eip.Spec.NodeSelector)
+	if err != nil {
+		return err
+	}
+	nodes, err := c.nodeLister.List(nodeSel)
+	if err != nil {
+		return err
+	}
+	c.consistentHashMap[eip.Name] = newNodeConsistentHashMap()
+	aliveNodes := c.aliveNodeList()
+	// node alive and node labels matches with externalIPPool nodeSelector
+	aliveAndMatchedNodes := sets.NewString()
```

Comment:
Neither NodeLister nor the memberlist members produces duplicate nodes, we should just use slice here as sets doesn't help here and you need to convert it to slice when using it.

Reply:
done